finetune: SGD optimizer, more CLI args

graehl opened the pull request:
add to ggml-opt learning rate (adamw alpha) cmdline arg, and an optimizer enum defaulting to adamw,
preparatory to work to support SGD

these are in common args a set of optimizer options active only for the new FINETUNE example (which includes all the previous finetune.cpp PERPLEXITY options as a precaution)

perhaps breaking with precedent, the ggml_opt_optimizer_params struct is included directly as args - if desired, we can instead just add learning rate and optimizer type to a struct independent of ggml-opt.h

as proposed in
https://github.com/ggml-org/llama.cpp/discussions/13835

graehl commented:
perhaps no need to review until i have an actual SGD impl in a follow-on, @JohannesGaessler - but a few general questions about contributing:

1. is it ok to make small retouches to ggml/ sources in this (llama.cpp) project with the expectation of getting the changes into the actual ggml repo later? are there any plans to submodule a ggml-in-llama branch to keep things straight(er)? 
2. is what i've got hee the expected way to add example-specific command line arguments? for finetune we definitely at least want to be able to vary the learning rate, which was formerly hard-coded.
3. were the PERPLEXITY args which i blindly added to the new FINETUNE example actually doing anything interesting? perhaps some should be dropped from finetune.
4. could you direct me to a .clang-format style file that might save me from accidentally re-indenting? i know i can set up clang-format to operate only on regions i've already changed ...

WilliamTambellini reviewed (approved):
you should better keep that change as it time to get more feedbacks/approval.

WilliamTambellini commented:
@ggerganov

JohannesGaessler reviewed (commented):
None of the previous perplexity-specific arguments are needed.

JohannesGaessler commented:
For adding an SDG optimizer, add a new ggml op like `OPT_STEP_SDG`. Add a CPU implementation as a fallback for any backend without an implementation. Add a CUDA implementation since that is (I assume) the backend which you intend to use in production. Add a test to `tests/test_backend_ops.cpp` to assert that the CPU and CUDA backends produce consistent results. Extend `ggml-opt.cpp` to conditionally use the new SDG optimizer step, condition the allocation of the optimizer momenta on the optimizer type.

graehl commented on common/arg.cpp:
```diff
@@ -2117,70 +2121,88 @@ common_params_context common_params_parser_init(common_params & params, llama_ex
+            throw std::invalid_argument("invalid --optimizer N (try 0)");
+          params.optimize.optimizer = (enum ggml_opt_optimizer)N;
+        }
+    ).set_examples({LLAMA_EXAMPLE_FINETUNE}));
```
will do (and removing ppx specific options)

JohannesGaessler commented:
Either way would be fine with me as long as there are at no point broken or unfinished features on master.

matiaslin reviewed (commented):
Looking forward to the next PR(s).

graehl commented:
you should see frivolous clang-format changes (using the project's .clang-format) only on lines changed in the PR (using git-clang-format). if there's something undesireable we could figure out what in the format config does it

JohannesGaessler commented:
Don't autoformat code en masse unless it's done in a dedicated PR, it makes it unnecessarily difficult to track what was actually changed in a PR.

JohannesGaessler commented:
Sorry, I didn't read the 

>only on lines changed in the PR

part.

graehl commented:
Second (actual usable SGD) commit is https://github.com/ggml-org/llama.cpp/pull/13873/commits/48a16bf250c04ee8fc7ea2f1cfb49c384b2a09a2 (also shows above here)

WilliamTambellini reviewed (changes requested):
Mix up different projects: change of CLI/renaming and SGD.  Need to split in 2 PRs.
@slaren  ?

graehl commented on ggml/src/ggml-opt.cpp:
```diff
@@ -770,7 +814,7 @@ void ggml_opt_eval(ggml_opt_context_t opt_ctx, ggml_opt_result_t result) {
         const float beta1h = 1.0f/(1.0f - powf(opt_pars.adamw.beta1, opt_ctx->iter));
         const float beta2h = 1.0f/(1.0f - powf(opt_pars.adamw.beta2, opt_ctx->iter));
-
+        const float keep           = 1.0f - opt_pars.adamw.alpha * opt_pars.adamw.wd;
```
I agree that it's not *likely* to matter, but it's 1. per parameter per epoch (ok, does seem unimportant now that I think further) and 2. i'm not confident cuda CC optimizes this and was hoping to learn more - would seem possible that w/o this we're loading repeatedly two floats instead of one - and mostly 3. this is exactly following precedent established for beta1h and beta2h, which are stored in the tensor just as i stored this quantity.

Anyway, totally willing, just curious what you think about the existing practice of saving beta1h and beta2h in light of this opinion that we're not compute bound.

graehl replied:
i checked it out - doesn't seem to change runtime noticeably as you predicted

graehl commented on common/arg.cpp:
```diff
@@ -1085,48 +1085,47 @@ static void common_params_print_completion(common_params_context & ctx_arg) {
+                                          "llama-batched-bench",
+                                          "llama-bench",
+                                          "llama-cli",
+                                          "llama-convert-llama2c-to-ggml",
+                                          "llama-cvector-generator",
+                                          "llama-embedding",
+                                          "llama-eval-callback",
+                                          "llama-export-lora",
+                                          "llama-finetune",
+                                          "llama-gen-docs",
+                                          "llama-gguf",
+                                          "llama-gguf-hash",
+                                          "llama-gguf-split",
+                                          "llama-gritlm",
+                                          "llama-imatrix",
+                                          "llama-infill",
+                                          "llama-mtmd-cli",
+                                          "llama-llava-clip-quantize-cli",
+                                          "llama-lookahead",
+                                          "llama-lookup",
+                                          "llama-lookup-create",
+                                          "llama-lookup-merge",
+                                          "llama-lookup-stats",
+                                          "llama-parallel",
+                                          "llama-passkey",
+                                          "llama-perplexity",
+                                          "llama-q8dot",
+                                          "llama-quantize",
+                                          "llama-qwen2vl-cli",
+                                          "llama-retrieval",
+                                          "llama-run",
+                                          "llama-save-load-state",
+                                          "llama-server",
+                                          "llama-simple",
+                                          "llama-simple-chat",
+                                          "llama-speculative",
+                                          "llama-speculative-simple",
+                                          "llama-tokenize",
+                                          "llama-tts",
+                                          "llama-vdot" };
```
i will try. recall that when i add something to a line you get clang-format using the project style file

graehl commented on ggml/include/ggml-opt.h:
```diff
@@ -74,16 +74,33 @@ extern "C" {
+
+    // "adamw" or "sgd" (case insensitive)
+    GGML_API const char *            ggml_opt_optimizer_name(enum ggml_opt_optimizer);
+    GGML_API enum ggml_opt_optimizer named_ggml_opt_optimizer(const char *);
```
done

graehl commented on ggml/include/ggml-opt.h:
```diff
@@ -74,16 +74,33 @@ extern "C" {
         struct {
             float alpha; // learning rate
-            float beta1;
-            float beta2;
+            float beta1;  // adamw
+            float beta2;  // adamw
             float eps;   // epsilon for numerical stability
-            float wd;    // weight decay for AdamW, use 0.0f to disable
+            float wd;    // weight decay for SGD or AdamW, use 0.0f to disable
         } adamw;
```
suppose we duplicate alpha, lr
then how do we make a single CLI arg pointing to both duplicates?
it seems we have to remember to manually copy to the linked places in our finetune.cpp example.

i do think it would be great to have clarity on which args apply to which optimizer method. maybe separation is the only way to achieve this or maybe we can mark a common set {alpha, wd}

graehl commented on ggml/include/ggml-opt.h:
```diff
@@ -74,16 +74,33 @@ extern "C" {
+        enum ggml_opt_optimizer optimizer;
+
+        // affects finetune.cpp only so far:
+        unsigned                epochs;  // max # of epochs sampling over training data
```
ok, suppose optimizer convergence criteria (how long to run) should go elsewhere. where do you recommend?

graehl commented on ggml/include/ggml-opt.h:
```diff
@@ -114,6 +131,8 @@ extern "C" {
 
         ggml_opt_get_optimizer_params get_opt_pars; // callback for calculating optimizer parameters
         void * get_opt_pars_ud;                     // userdata for calculating optimizer parameters
+        struct ggml_opt_optimizer_params
+            opt_params;  // holds result of get_opt_pars(get_opt_pars_ud) after ggml_opt_init (could call get_opt_pars repeatedly instead)
```
i can easily remove it, of course.
but currently get_opt_pars is called twice per epoch.
this feels wrong.

JohannesGaessler commented on ggml/include/ggml.h:
```diff
@@ -2063,6 +2064,12 @@ extern "C" {
             struct ggml_tensor  * adamw_params); // parameters such a the learning rate
 
+    // SGD (with weight decay) step
+    GGML_API struct ggml_tensor * ggml_opt_step_sgd(struct ggml_context * ctx, struct ggml_tensor * a,
+                                                    struct ggml_tensor * grad,
+                                                    // parameters: alpha (learning rate), wd (weight decay)
+                                                    struct ggml_tensor * adamw_params);
```
Please make the indentation consistent with the other functions.

JohannesGaessler commented on ggml/src/ggml-opt.cpp:
```diff
@@ -223,15 +227,36 @@ struct ggml_opt_optimizer_params ggml_opt_get_default_optimizer_params(void * us
 
     ggml_opt_optimizer_params result;
 
-    result.adamw.alpha = 0.001f;
+    result.common.alpha = 1e-8f;
```
Don't change the default, it's the value that works well for the ggml MNIST example and since there is no universally correct value for the learning rate I've chosen to make the default consistent with PyTorch. Change the default set by the more specialized code in llama.cpp instead.

graehl replied:
ok

JohannesGaessler commented on ggml/src/ggml.c:
```diff
@@ -5184,6 +5184,27 @@ struct ggml_tensor * ggml_opt_step_adamw(
+    return result;
+}
+
+// opt_step_sgd
+
```
```suggestion
```

JohannesGaessler commented on ggml/src/ggml.c:
```diff
@@ -5184,6 +5184,27 @@ struct ggml_tensor * ggml_opt_step_adamw(
     return result;
 }
 
+// opt_step_adamw
```
```suggestion
// opt_step_sgd
```

graehl replied:
oh, i see :) i did this. i thought i was copying previous style

JohannesGaessler commented on tests/test-backend-ops.cpp:
```diff
@@ -763,7 +763,8 @@ struct test_case {
         ggml_tensor * out = build_graph(ctx.get());
 
-        if ((op_name != nullptr && op_desc(out) != op_name) || out->op == GGML_OP_OPT_STEP_ADAMW) {
+        if ((op_name != nullptr && op_desc(out) != op_name) || out->op == GGML_OP_OPT_STEP_ADAMW ||
+            out->op == GGML_OP_OPT_STEP_SGD) {
```
```suggestion
        if ((op_name != nullptr && op_desc(out) != op_name) || out->op == GGML_OP_OPT_STEP_ADAMW ||
                out->op == GGML_OP_OPT_STEP_SGD) {
```

The identation should either be visually distinct from the line below or there should be a blank line separating them.

graehl replied:
ok

JohannesGaessler reviewed (commented):
Unless it's something very minor, please let me resolve conversations since that is how I like to track the TODOs (from my side) in a PR.

graehl commented on ggml/src/ggml-opt.cpp:
```diff
@@ -492,23 +529,31 @@ static void ggml_opt_build(ggml_opt_context_t opt_ctx) {
-            ggml_set_name(opt_step, (std::string("AdamW step for ") + std::string(node->name)).c_str());
-
+            struct ggml_tensor * opt_step;
+            if (moment) {
```
I understand this in the abstract but have no idea what concrete change you're requesting if any beyond what I have now -  what mechanism do you think is worth implementing now for this goal of supporting N optimizers in the future? Do you want to see a new free fn returning a bool? Do you want a new 'optimizer' interface with methods (I don't see any sort of precedent for C++ interfaces in ggml, as opposed to llama, so presumably you'd be asking for a C style function-pointers thing). Specify it and I can do it.

graehl commented on common/arg.cpp:
copy re: resolve.
as i said, the intention is to autoformat only the new code i add. if i accidentally changed other lines and they were affected, i'm happy to revert

graehl commented on .clang-format:
i was attempting to comply with your request to not change the formatting of a long initializer list of string literals. in my view the best way to reduce formatting make-work is to have the .clang-format match the codebase better. can revert

graehl commented on examples/training/finetune.cpp:
```diff
@@ -60,8 +59,17 @@ int main(int argc, char ** argv) {
-    struct ggml_opt_optimizer_params optimizer_params = ggml_opt_get_default_optimizer_params(nullptr);
-    optimizer_params.adamw.alpha = 1e-7f; // learning rate
+    struct ggml_opt_optimizer_params & optimizer_params = params.optimize;
+    if (optimizer_params.optimizer == GGML_OPT_OPTIMIZER_SGD) {
+        double was = (double) optimizer_params.common.alpha;
+        double by  = 1e2;
+        double to  = was * by;
+        LOG_INF("sgd multiplying -lr by %.3g (no momentum) from -lr: %.2g to %.2g\n", by, was, to);
+        optimizer_params.common.alpha = to;
+    }
```
do you want me to distinguish default vs explicit -lr so we can have sensible defaults for both methods, or are you just wanting this removed, period?

graehl commented:
re: 'I think it's more important to use the exact same data layout for the structs in ggml-opt.cpp and the tensors for the corresponding operations defined in ggml.h than to deduplicate a few of the parameters.'

There is something slightly strange IMO in relying on "at position 4 you will find the weight decay, just as it is in this ggml-opt.h adamw struct". The part that feels off specifically is the use of the raw constant 4 in the backends, as opposed to some kind of cast to struct * that makes the requirement legible (see, we just memcpy, it has to be the same layout), or the use of a named integer constant/enum i_adamw_wd = 4, at which point layout shouldn't matter.

To keep the layout correspondence you want, it seems i have to drop the 'common' idea for keeping adamw params out of sgd. I'm happy to do this (common was in response to what i thought was your request) - please confirm.

JohannesGaessler commented:
>There is something slightly strange IMO in relying on "at position 4 you will find the weight decay, just as it is in this ggml-opt.h adamw struct". The part that feels off specifically is the use of the raw constant 4 in the backends, as opposed to some kind of cast to struct * that makes the requirement legible (see, we just memcpy, it has to be the same layout), or the use of a named integer constant/enum i_adamw_wd = 4, at which point layout shouldn't matter.

It would be completely fine for you to also define structs that provide a safer way to retrieve the optimizer parameters. The scenario that is relevant for me is this: I have a debugger open and want to inspect the contents of some tensor. It is more convenient for me when I can inspect the memory of the tensor by simply casting it to float vs. having to cast it to some specific struct. If the memory layout is inconsistent across multiple points in the program that adds an additional thing that I have to keep track of during debugging (which is usually the most time-consuming part of my work).

>To keep the layout correspondence you want, it seems i have to drop the 'common' idea for keeping adamw params out of sgd. I'm happy to do this (common was in response to what i thought was your request) - please confirm.

Yes, please do that.

graehl commented:
any update? the memory savings are good and the behavior is what you'd expect from SGD

graehl commented:
Sorry - didn't realize one of my merges somehow resulted in touching almost every line of arg.cpp - fixed w/ a rebase. Is there anything blocking this now?

WilliamTambellini commented:
@JohannesGaessler  any comment?

JohannesGaessler commented on common/arg.cpp:
```diff
@@ -3376,5 +3376,34 @@ common_params_context common_params_parser_init(common_params & params, llama_ex
+        [](common_params & params, const std::string & name) {
+            params.optimize.optimizer = ggml_opt_get_optimizer(name.c_str());
+            if (params.optimize.optimizer == GGML_OPT_OPTIMIZER_COUNT) {
+                throw std::invalid_argument("invalid --optimizer (try adamw)");
```
```suggestion
                throw std::invalid_argument("invalid --optimizer, valid options: adamw, sgd");
```

graehl replied:
since my initial attempt at passing a host float to .cu by value failed (i think you have to actually load it via a cuda float*), i'll just drop 'keep' precalc. as you pointed out it's, though maybe saving CO2 emissions or something, not faster in real time.

JohannesGaessler commented on ggml/include/ggml-opt.h:
```diff
@@ -74,16 +74,30 @@ extern "C" {
         } adamw;
+
+        // only GGML_OPT_OPTIMIZER_ADAMW allocates m, v per parameter
+        enum ggml_opt_optimizer_type optimizer;
```
Remove this from `ggml_opt_optimizer_params` and add it to `ggml_opt_params` instead. The optimizer parameters are fetched once per epoch and having this property here would suggest that it's possible to change the optimizer type during training. But that is going to make the code much more complicated and IIRC we agreed that we would not add support for it.

JohannesGaessler commented on ggml/src/ggml-opt.cpp:
```diff
@@ -312,12 +337,27 @@ static ggml_cgraph * dup_graph(ggml_context * ctx, ggml_cgraph * src) {
 
+    struct ggml_opt_optimizer_params const & opt_pars = opt_ctx->opt_pars();
+    const bool                               moment =
+        // or has_moment is per https://github.com/ggml-org/llama.cpp/pull/13873#discussion_r2116366095
+        // - do not allow sgd epochs after adamw
+        // (which in the current impl would require clearing the previously n_nodes-sized grad_m and grad_v)
+        opt_ctx->build_type_alloc == GGML_OPT_BUILD_TYPE_OPT && (need_moment(opt_pars) || has_moment(opt_ctx));
+
     const bool accumulate = opt_ctx->build_type_alloc >= GGML_OPT_BUILD_TYPE_GRAD &&
-        !(opt_ctx->static_graphs && opt_ctx->build_type_alloc == GGML_OPT_BUILD_TYPE_OPT && opt_ctx->opt_period == 1);
+                            !(opt_ctx->static_graphs && moment && opt_ctx->opt_period == 1);
```
The optimizer type should be constant for the lifetime of `ggml_opt_context` so it's enough to just check whether the constant optimizer type needs the momenta.

graehl replied:
right, i can agree with all that. still unclear on exactly what you want changed. feel free to change what you like as long as it still works. probably you can do that with more confidence than me.

JohannesGaessler replied:
Add the optimizer type to `ggml_opt_params` and `ggml_opt_context`, set the value of the context during initialization, do not change it afterwards, use it for the `need_moment` logic. Does that make it clear enough?

graehl replied:
no, those instructions don't parse. i have already put the optimizer type in context. can you refer to the latest push where we have already:

    const enum ggml_opt_optimizer_type optimizer = opt_ctx->optimizer;
    ...
    const bool need_moment = optimizer == GGML_OPT_OPTIMIZER_TYPE_ADAMW;

it might be that i already did what you wanted, perhaps.

JohannesGaessler commented on ggml/src/ggml-opt.cpp:
```diff
@@ -492,23 +531,41 @@ static void ggml_opt_build(ggml_opt_context_t opt_ctx) {
-            ggml_set_name(v,        (std::string("AdamW v for ")    + std::string(node->name)).c_str());
-            ggml_set_name(opt_step, (std::string("AdamW step for ") + std::string(node->name)).c_str());
-
+            struct ggml_tensor * m = 0;
+            struct ggml_tensor * v = 0;
```
```suggestion
            struct ggml_tensor * m = nullptr;
            struct ggml_tensor * v = nullptr;
```

JohannesGaessler commented on ggml/src/ggml-opt.cpp:
```diff
@@ -492,23 +531,41 @@ static void ggml_opt_build(ggml_opt_context_t opt_ctx) {
+                    GGML_ABORT("fatal error");
+                    break;
+            }
+            step_prefix.resize(
+                step_prefix_len_orig);  // to avoid recreating a new step_prefix string temp n_nodes times
+            ggml_set_name(opt_step, (step_prefix += node->name).c_str());
```
```suggestion
            step_prefix.resize(step_prefix_len_orig);  // to avoid recreating a new step_prefix string temp n_nodes times
            step_prefix += node->name;
            ggml_set_name(opt_step, step_prefix.c_str());
```

JohannesGaessler commented on ggml/src/ggml-opt.cpp:
```diff
@@ -756,29 +812,29 @@ void ggml_opt_alloc(ggml_opt_context_t opt_ctx, bool backward) {
+        const float keep           = 1.0f - opt_pars.adamw.alpha * opt_pars.adamw.wd;
         float * adamw_par_data = ggml_get_data_f32(opt_ctx->adamw_params);
-        adamw_par_data[0] = opt_pars.adamw.alpha;
+        adamw_par_data[0]          = opt_pars.adamw.alpha;
```
```suggestion
        adamw_par_data[0] = opt_pars.adamw.alpha;
```

JohannesGaessler commented on tests/test-opt.cpp:
```diff
@@ -40,7 +40,7 @@ struct helper_ctx_data {
 static ggml_opt_optimizer_params helper_get_test_opt_pars(void * userdata) {
     ggml_opt_optimizer_params result = ggml_opt_get_default_optimizer_params(userdata);
-    result.adamw.alpha = 1.0f;
+    result.adamw.alpha               = 1.0f;
```
```suggestion
    result.adamw.alpha = 1.0f;
```

JohannesGaessler commented on tests/test-opt.cpp:
```diff
@@ -687,7 +687,7 @@ static std::pair<int, int> test_gradient_accumulation(
 static ggml_opt_optimizer_params helper_get_regression_opt_pars(void * userdata) {
     ggml_opt_optimizer_params result = ggml_opt_get_default_optimizer_params(userdata);
-    result.adamw.alpha = 0.1f;
+    result.adamw.alpha               = 0.1f;
```
```suggestion
    result.adamw.alpha = 0.1f;
```

JohannesGaessler reviewed (commented):
One issue that remains is the memory layout for the parameters for the optimizer steps. They need to be compact, human-readable, and limited to those parameters that the optimizer step actually uses. In practical terms that the AdamW optimizer step should use the same memory layout as on master. The SGD optimizer step should only be using alpha and the weight decay. The layout of `ggml_opt_optimizer_params` should mirror the layout of the ggml ops. Add a new sub-struct `sgd` with alpha and the weight decay. Do not re-use the adamw struct for SGD.

Remove the parameter `keep`. As I said, if you want to do this optimization the correct way to do it is to calculate this parameter in the ggml ops themselves. With CUDA it can be calculated once in host code and then passed to device code.

graehl commented:
ok, that was a lot of changes, i believe i got them all and it looks like both adamw and sgd still work. i did have one question re: has_moment

JohannesGaessler commented on ggml/src/ggml-opt.cpp:
```diff
@@ -316,8 +348,14 @@ static void ggml_opt_build(ggml_opt_context_t opt_ctx) {
     GGML_ASSERT(opt_ctx->ctx_compute && "no compute context set, either use static graphs or set one with ggml_opt_prepare_alloc");
     GGML_ASSERT((!opt_ctx->static_graphs || opt_ctx->inputs->data) && "when using static graphs the inputs must be allocated statically");
 
+    const enum ggml_opt_optimizer_type optimizer = opt_ctx->optimizer;
+    const bool has_moment = !opt_ctx->grad_m.empty();
+    const bool need_moment = optimizer == GGML_OPT_OPTIMIZER_TYPE_ADAMW;
+    const bool moment =
+        opt_ctx->build_type_alloc == GGML_OPT_BUILD_TYPE_OPT && (need_moment || has_moment);
+
```
```suggestion
    const bool moment = opt_ctx->build_type_alloc == GGML_OPT_BUILD_TYPE_OPT &&  opt_ctx->optimizer == GGML_OPT_OPTIMIZER_TYPE_ADAMW;
```

Sorry, I didn't look at the code again before writing my last comment. This is what I meant the logic to be.

graehl commented:
> One more thing: at some point the unit tests for ggml-opt were commented out in `tests/CMakeLists.txt`. Please adapt them to work with the new code (should be relatively easy). Adding tests for SGD would be appreciated but is not required for me approving this PR.

sure, no problem, should we push with tests re-disabled? comment says they're very slow

graehl committed:
examples/finetune -opt SGD (stochastic gradient descent) memory opt

add unit tested GGML_OPT_OPTIMIZER_SGD to ggml - avoids allocating
m, v tensors.

support finetune.cpp arg -opt SGD (or sgd). (default adamw as before)

llama 3.2-1b-F32 result: observed 11gb gpu ram (41 sec/epoch)
when using SGD instead of 19gb (55 sec/epoch) using adamw.
(wikipedia 100 lines finetune)

(
using the same GPU memory, adamw can only do before OOM 512
batch/context, reaching:
train: [███████▉] data=0000140/0000140 loss=0.02575±0.00099 acc=99.52±0.03% t=00:00:47 ETA=00:00:00
val:   [███████▉] data=0000008/0000008 loss=4.76565±0.28810 acc=41.46±0.77% t=00:00:00 ETA=00:00:00

SGD is superior, though it converges slower, with max before OOM 1728
batch/context (esp see the better validation perf):
train: [███████▉] data=0000039/0000039 loss=0.00371±0.00010 acc=99.96±0.01% t=00:00:41 ETA=00:00:00
val:   [███████▉] data=0000003/0000003 loss=5.11406±0.76034 acc=48.01±0.69% t=00:00:01 ETA=00:00:00
)

note: when finetuning long enough (or w/ enough -lr),
validation accuracy *eventually* drops ('catastrophic forgetting')

-lr-half (halflife) option useful for SGD to avoid oscillation or
super slow underdamped learning (makes setting -lr more forgiving).
terminal -lr for now is set by lr-halvings i.e. if you want at most
1/8 the inital -lr you set -lr-halvings 3.

note: objective loss not directly comparable between adamw, sgd? -
check perplexity or accuracy or consider relative improvements
for convergence

new finetune args -wd 1e-9 to enable weight decay in sgd or adamw,
and max -epochs N (default 2 as before)

cache (1 - wd*alpha) in 'adamw' opt struct -
no noticeable perf benefit, disabled (still done
for new SGD though)

since opt. memory is pre-allocated, the ggml_opt_get_optimizer_params
would probably be able to change between SGD and AdamW with each epoch
but would need to use adamw for the first (unconfirmed - no cmdline arg
to set such a policy yet)

test-opt checks adamw as before and now sgd (except for a few disabled
tests for sgd only; probably just needs logging values and adding
alternate reference values);  tolerance on the 'regression'
test is broader for sgd (so we don't need many more epochs)

JohannesGaessler reviewed (commented):
What is your intent with this PR? If you want to move towards getting it merged you should stop adding new features. From my end it doesn't really matter either way since there is to my knowledge no one else working on this code in parallel.

graehl commented on common/common.h:
```diff
@@ -219,6 +222,25 @@ enum common_reasoning_format {
     COMMON_REASONING_FORMAT_DEEPSEEK,        // Extract thinking tag contents and return as `message.reasoning_content`, including in streaming deltas.
 };
 
+struct lr_decay {
+    float lr              = 1e-5;
+    float halflife_epochs = 100;
+    float halvings        = 10;
+
+    float decayed(float epoch) const {
+        float maxepoch = halvings * halflife_epochs;
+        return lr * std::pow(.5, (epoch > maxepoch ? maxepoch : epoch) / halflife_epochs);
+    }
+};
+
+struct lr_opt : lr_decay {
+  float epoch = 0;
+  float wd = 0;
+  unsigned epochs = 2;
+};
+
+struct ggml_opt_optimizer_params common_lr_opt_pars(void * userdata);
```
It was intentional to have a step only at the end of a full epoch; I realize this is not how everyone does it.
I don't agree that discrete factors of 2 follow, halflife is a configurable (possibly fractional) # of epochs.

graehl replied:
if the proposal is to allow gradual decay within an epoch perhaps this can come in a separate patch incl all sorts of decay policies (incl specifying initial/terminal alpha as you prefer). obviously the way it's done now works without touching ggml but can only update after a whole epoch, which it sounds like you don't love (i think you wouldn't really *want* to vary LR during an epoch unless you had random shuffling or sampling w/ replacement - is that currently happening?)

graehl replied:
i do think it's a good feature *provided you have sampling or shuffling* - should overall train effectively in less time w/ steady decay IMO

graehl replied:
oh hey, to be clear my use of float is safe - epoch is not a float when it's incremented (see finetune.cpp `    for (unsigned epoch = 0; epoch < lr.epochs; ++epoch) {`, and i totally agree. if i recall correctly you lose the ability to increment somewhere in the millions :)

graehl commented on common/arg.cpp:
```diff
@@ -3376,5 +3376,56 @@ common_params_context common_params_parser_init(common_params & params, llama_ex
+                   [](common_params & params, const std::string & value) { params.lr.lr = std::stof(value); })
+            .set_examples({ LLAMA_EXAMPLE_FINETUNE }));
+    add_opt(common_arg(
+                { "-lr-half", "--learning-rate-halflife-epochs" }, "N",
+                string_format("reduce lr in half every N epochs (default: %.3g)", (double) params.lr.halflife_epochs),
+                [](common_params & params, const std::string & value) { params.lr.halflife_epochs = std::stof(value); })
+                .set_examples({ LLAMA_EXAMPLE_FINETUNE }));
+    add_opt(common_arg({ "-lr-halvings", "--learning-rate-halvings" }, "N",
+                       string_format("max N lr halvings (default: %.3g)", (double) params.lr.halvings),
+                       [](common_params & params, const std::string & value) { params.lr.halvings = std::stof(value); })
+                .set_examples({ LLAMA_EXAMPLE_FINETUNE }));
```
Sure, but then you need to distinguish one being explicitly specified vs not. i.e. mine can be left at default while tweaking just -lr
Not a big deal, lmk

graehl replied:
with the lr-halvings you have the main useful thing you get w/ an absolute final LR - the ability to just specify more epochs without accidentally getting an ineffective ~0 learing rate. would be fine to add the option you prefer in addition at some point

graehl commented:
Let's merge, I'll work on just the changes you need, stand by for possibly last draft. Want opt_period removed?

graehl commented on ggml/src/ggml-opt.cpp:
```diff
@@ -223,15 +230,39 @@ struct ggml_opt_optimizer_params ggml_opt_get_default_optimizer_params(void * us
     return result;
 }
 
+GGML_API const char * ggml_opt_optimizer_name(enum ggml_opt_optimizer_type o) {
+    switch (o) {
+        case GGML_OPT_OPTIMIZER_TYPE_ADAMW:
+            return "adamw";
+        case GGML_OPT_OPTIMIZER_TYPE_SGD:
+            return "sgd";
+        default:
+            return "undefined";
+    };
+}
+
+GGML_API enum ggml_opt_optimizer_type ggml_opt_get_optimizer(const char * n) {
+    if (!strcasecmp("adamw", n)) {
+        return GGML_OPT_OPTIMIZER_TYPE_ADAMW;
+    } else if (!strcasecmp("sgd", n)) {
+        return GGML_OPT_OPTIMIZER_TYPE_SGD;
+    } else {
+        return GGML_OPT_OPTIMIZER_TYPE_COUNT;
+    }
+}
```
wouldn't this make them unavailable to someone using bare ggml (no llama)?

graehl replied:
ok, easy change

graehl commented on ggml/src/ggml-opt.cpp:
```diff
@@ -223,15 +230,39 @@ struct ggml_opt_optimizer_params ggml_opt_get_default_optimizer_params(void * us
-    result.adamw.wd    = 0.0f;
+    result.adamw.wd     = 0.0f;
+
+    result.sgd.alpha = 1e-6;
+    result.sgd.wd = 1e-10;
```
(changed per req)

JohannesGaessler replied:
Defaults for language models in particular should be set in llama.cpp, not ggml.

JohannesGaessler commented:
>Want opt_period removed?

I want it removed as a CLI argument, the parameterization should be done via `--batch` and `--ubatch` as is done on master.

JohannesGaessler commented on common/arg.cpp:
```diff
@@ -3376,5 +3376,56 @@ common_params_context common_params_parser_init(common_params & params, llama_ex
+                       string_format("optimizer max # of epochs (default: %d)", params.lr.epochs),
+                       [](common_params & params, int epochs) { params.lr.epochs = epochs; })
+                .set_examples({ LLAMA_EXAMPLE_FINETUNE }));
+    add_opt(common_arg({ "-period", "--opt-period" }, "N",
+                       string_format("make logical batch this multiple of physical batch - needs more memory for accumulation if >1 (default: %d)", params.opt_period),
+                       [](common_params & params, int opt_period) { params.opt_period = opt_period; })
+                .set_examples({ LLAMA_EXAMPLE_FINETUNE }));
```
Revert the change.

graehl commented:
> > > Want opt_period removed?
> > 
> > 
> > I want it removed as a CLI argument, the parameterization should be done via `--batch` and `--ubatch` as is done on master.
> 
> ok, no problem. keep in mind that the parameterization w/ those args has no effect because of code that chops off batch to match ubatch. do you want me to open it up? it seems erroneous to limit it thus and opt_period can be computed (assuming physical batch divides logical evenly). otherwise i just kill the option i was using to get opt_period and save it for another day

`void llama_context::opt_epoch_iter(
        ...) {
    const uint32_t n_ctx    = llama_model_n_ctx_train(&model);
    const uint32_t n_batch  = std::min(this->n_batch(),  n_ctx); // logical batch limited to no more than n_ctx
    const uint32_t n_ubatch = std::min(this->n_ubatch(), n_batch);
   // physical batch limited to no more than (already limited) logical batch
`
i suppose in my testing before i didn't realize n_ctx was what was stopping me from having a large batch size. anyway, removing the opt_period arg.

JohannesGaessler commented on common/common.cpp:
```diff
@@ -1535,3 +1535,32 @@ ggml_opt_dataset_t common_opt_dataset_init(struct llama_context * ctx, const std
+    return result;
+}
+
+GGML_API const char * ggml_opt_optimizer_name(enum ggml_opt_optimizer_type o) {
+    switch (o) {
+        case GGML_OPT_OPTIMIZER_TYPE_ADAMW:
+            return "adamw";
+        case GGML_OPT_OPTIMIZER_TYPE_SGD:
+            return "sgd";
+        default:
+            return "undefined";
+    };
+}
```
I thought about the issue of where to put which code some more. The mapping from string to optimizer type should be in `common.cpp` since I think parsing CLI arguments is the responsibility of the "user code". But the mapping in the other direction, meaning this function `ggml_opt_optimizer_name` that maps an optimizer type to a string, should be in `ggml-opt.cpp` after all, just like there are equivalent functions in `ggml.c`.

JohannesGaessler reviewed (commented):
Prior to my review there were already unresolved requests for changes that will need to be addressed prior to merging.

graehl commented on common/arg.cpp:
```diff
@@ -3376,5 +3376,52 @@ common_params_context common_params_parser_init(common_params & params, llama_ex
+            .set_examples({ LLAMA_EXAMPLE_FINETUNE }));
+    add_opt(common_arg(
+                { "-lr-half", "--learning-rate-halflife-epochs" }, "N",
+                string_format("reduce lr in half every N epochs (default: %.3g)", (double) params.lr.halflife_epochs),
```
suggestion is incorrect but i will reword to avoid the "N" which i guess you don't like

graehl commented on common/arg.cpp:
```diff
@@ -3376,5 +3376,52 @@ common_params_context common_params_parser_init(common_params & params, llama_ex
+                   [](common_params & params, const std::string & value) { params.lr.lr = std::stof(value); })
+            .set_examples({ LLAMA_EXAMPLE_FINETUNE }));
+    add_opt(common_arg(
+                { "-lr-half", "--learning-rate-halflife-epochs" }, "N",
```
i think this is worse, but ok

graehl commented:
if you prefer lr-min enough to do it yourself, great. otherwise i'll decline

JohannesGaessler commented:
If we can't reach an agreement regarding the exact implementation regarding a decaying learning rate I would say to just strip the feature from this PR.

The other pre-existing, unresolved issues I mentioned were the ones in `ggml-opt.cpp`. If you go through the file on Github, you should still be seeing the comments. I will add additional comments if there is anything.

Sorry for being difficult, social skills are not my strong point.

graehl commented:
> If we can't reach an agreement regarding the exact implementation regarding a decaying learning rate I would say to just strip the feature from this PR.
> 
> The other pre-existing, unresolved issues I mentioned were the ones in `ggml-opt.cpp`. If you go through the file on Github, you should still be seeing the comments. I will add additional comments if there is anything.
^^^
I'll try to figure out how to do this.

> Sorry for being difficult, social skills are not my strong point.

same :)

graehl commented:
any update? all handled on my end afaict

JohannesGaessler commented:
Also there is still global state in the unit tests.

graehl commented:
pushed: removed 'half' related args, verify lr0/lr-min behavior, removed global var from test-opt (verified still passes)

please let me know if you find any other change you want as i'm keen to wrap this up

graehl commented:
and no objection to any cosmetics you want to slap on yourself, of course!

JohannesGaessler commented:
Yes, please go to "Files changed" on the Github interface and look at my comments on `ggml-opt.cpp`. The changes aren't big but I would like to assert that they are not causing issues.

graehl commented:
mea culpa - i didn't push after testing locally :)

graehl commented:
You asked me to enable test-opt but it doesn't free its resources and so some of the CI jobs (which run under valgrind/asan) fail. How do you want to proceed on those failures e.g. in test-opt calls to ggml_opt_fit

graehl commented:
> You asked me to enable test-opt but it doesn't free its resources and so some of the CI jobs (which run under valgrind/asan) fail. How do you want to proceed on those failures e.g. in test-opt calls to ggml_opt_fit

Any idea where I can find the CI logs? They seem to have vanished for me after I push -f

JohannesGaessler commented:
Sorry, I don't know. Make fixup commits since the PR will end up being squashed anyways?

graehl commented:
fixed: another unreachable break, windows strcasecmp, skip SGD test-opt case (keep ADAMW) for vulkan. should (nearly) pass CI now

lexasub commented:
@graehl need fix windows builds))

lexasub commented:
@graehl you may create new mr(from other branch) which willn't break windows build, then mr will small) (if fix windows is very hard)

graehl commented:
fixed windows dll GGML_API build
fixed test-backend-ops (avoid SGD * vulkan, still TODO)

lexasub commented:
![image](https://github.com/user-attachments/assets/4184718d-aa90-4052-b9ef-85a63b0c0715)
graehl:finelayer rebased on main (finetune.cpp)

graehl commented:
> ![image](https://private-user-images.githubusercontent.com/16270007/462229338-4184718d-aa90-4052-b9ef-85a63b0c0715.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE2MDIyMjQsIm5iZiI6MTc1MTYwMTkyNCwicGF0aCI6Ii8xNjI3MDAwNy80NjIyMjkzMzgtNDE4NDcxOGQtYWE5MC00MDUyLWI5ZWYtODVhNjNiMGMwNzE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzA0VDA0MDUyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2YmY5MmIyZmQ0MTRiNTI1MzRmZTZhNDkyZjgwY2I1YmQ0NThhZWM0OTExN2VjMmZlNGNjNGQ1MDFhZjI2MWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3Go3uFR7EHp1G1qwki0f8i078ox8TBFNNPdhAkFpIxo) graehl:finelayer rebased on master (finetune.cpp)

(I reverted the unintentional diffs shown)

graehl commented:
seems test-opt can't link on windows (test-opt.obj : error LNK2019: unresolved external symbol ggml_backend_is_cpu referenced in function main [D:\a\llama.cpp\llama.cpp\build\tests\test-opt.vcxproj]
test-opt.obj : error LNK2019: unresolved external symbol ggml_backend_cpu_set_n_threads referenced in function main [D:\a\llama.cpp\llama.cpp\build\tests\test-opt.vcxproj]
D:\a\llama.cpp\llama.cpp\build\bin\Release\test-opt.exe : fatal error LNK1120: 2 unresolved externals [D:\a\llama.cpp\llama.cpp\build\tests\test-opt.vcxproj])
might need to resolve by re-disabling the test on windows; any suggestion to fix otherwise?

graehl commented:
do we care about the vulkan test-backend-ops timing out? seems nothing is actually wrong ...
31:   FLASH_ATTN_EXT(hsk=576,hsv=512,nh=4,nr23=[1,1],kv=512,nb=3,mask=1,max_bias=0.000000,logit_softcap=0.000000,prec=f32,type_KV=q8_0,permute=[0,1,2,3]): OK
31:   FLASH_ATTN_EXT(hsk=576,hsv=512,nh=4,nr23=[1,1],kv=512,nb=3,mask=1,max_bias=0.000000,logit_softcap=0.000000,prec=f32,type_KV=q8_0,permute=[0,2,1,3]): OK
31/36 Test #31: test-backend-ops ..................***Timeout 3600.07 sec
test 34

JohannesGaessler commented on tests/test-opt.cpp:
```diff
@@ -851,54 +908,58 @@ int main(void) {
-        printf("\n");
+            char const* devname = ggml_backend_dev_name(devs[i]);
+            printf("Backend %zu/%zu: %s\n", i + 1, dev_count, devname);
+            if (optim == GGML_OPT_OPTIMIZER_TYPE_SGD && !strcmp(devname, "Vulkan0"))
+              // TODO
+              continue;
```
```suggestion
```

graehl replied:
recall: this test was disabled previously
now: it doesn't work in vulkan.
so, either test needs to not be run on vulkan
or this needs to stay

graehl commented:
The mac CI test seems unrelated (timeout - just increase it above 900s on this platform?) https://github.com/ggml-org/llama.cpp/actions/runs/16152514813/job/45670340308?pr=13873  27 - test-thread-safety (Timeout)                      main
27: 14.14.041.515 I Model 1/2, Context 2/4: The meaning of life is on a cold day. Jane and her mom are going to the store. Jane likes the store, but she has a problem.
27: "Mom, why are you so upset?" Jane's mom asked her.
27: "I'm upset because I'm out of sugar," Jane replied.
27: "Oh Jane, why don't we make some sugar? We will do it together," her mom suggested.
27: Jane was excited. She was ready to make sugar with her mom. But, before she started, her mom said, "Jane, why don't you use a jar of sugar
27: 
27/36 Test #27: test-thread-safety ................***Timeout 900.04 sec

graehl commented:
also presumably you're aware of this but the 'build-linux-cross' vulkan CI machines are misconfigured (at least intermittently) 
<img width="268" height="189" alt="image" src="https://github.com/user-attachments/assets/5b1c965a-09c0-4581-87b4-c1afc5f2acfa" />

lexasub commented:
Hi @graehl!
Apologies for the interruption, but I’m completely tapped out after trying to get Parquet dataset support working via gguf_context in llama.cpp. The code almost works, but my eyes are fried after a week of "a week of debugging hell" 🙃

What’s done:

Text, Parquet reading via gguf_context (loading datasets into tensors, etc.).
Converting it to custom gguf. 
Tested on whatever was handy (small test files, no serious datasets yet).
Code is a bit "hacky" in places, but logic seems solid.
Why I’m reaching out:

I might push more changes to my branch in the next couple of days (probably by tomorrow), so just a heads-up if you start refactoring.
Your fine-tuning MR looks cleaner than my spaghetti – want to integrate Parquet support into your example or review my mess?
Key question: Do you even want Parquet support in llama.cpp? If yes, let me know, and I’ll avoid breaking things further while you review.
Where to find my code:

Branch: [finetune-preview](https://github.com/lexasub/llama.cpp/tree/finetune-preview)
If you take this on – massive thanks in advance! 😊 If not, I’ll circle back later, but my brain needs a break for now.

0cc4m committed:
Vulkan: Implement GGML_OP_OPT_STEP_SGD

0cc4m committed:
tests: Fix OPT_STEP_SGD test-backend-ops

graehl commented:
`2025-07-21T10:37:03.7706475Z 29: /home/runner/work/llama.cpp/llama.cpp/ggml/src/ggml-backend.cpp:735: pre-allocated tensor (loss_sum) in a buffer (WebGPU) that cannot run the operation (SUM)`
What does this mean? Is it something I'm expected to fix? I could handle it by skipping the "WebGPU" backend in test-opt
As a reminder, this is happening not because of my code, but because test-opt used to be disabled in cmake.

JohannesGaessler commented:
>What does this mean? Is it something I'm expected to fix? I could handle it by skipping the "WebGPU" backend in test-opt
>As a reminder, this is happening not because of my code, but because test-opt used to be disabled in cmake.

Fine to either skip WebGPU or disable the tests again for this PR.

graehl commented:
pushed an attempt to just skip the WebGPU device but otherwise leave test-opt enabled. if that doesn't work will just comment out the cmake for test-opt as before

0cc4m commented:
@graehl I finished the SGD Vulkan implementation in https://github.com/graehl/llama.cpp/pull/1. It passes the test now. Let me know if you see an issue with it.

lexasub commented:
@JohannesGaessler completed for final review/merge

вторник, 5 августа 2025 г. пользователь Jonathan Graehl <
***@***.***> написал:

> *graehl* left a comment (ggml-org/llama.cpp#13873)
> <https://github.com/ggml-org/llama.cpp/pull/13873#issuecomment-3152649712>
>
> I'm not aware of anything I can do on my end to get this merged (is
> someone waiting on me that I'm unaware of?). I just marked a 'conflict'
> above resolved but I think ultimately the permission is with the llama.cpp
> maintainers
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ggml-org/llama.cpp/pull/13873#issuecomment-3152649712>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD4EFN4WQNYEI6UBSCIFP4L3L7PBXAVCNFSM6AAAAAB6DZ332KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCNJSGY2DSNZRGI>
> .
> You are receiving this because you commented.Message ID: <ggml-org/llama.
> ***@***.***>
>

0cc4m commented:
You also didn't address 2 crucial comments from Johannes:

<img width="1016" height="1395" alt="image" src="https://github.com/user-attachments/assets/63899fa1-fcd9-4281-88c9-f26456c5bfa2" />

This problem causes issues with running tests, that's why I also fixed it in my branch. If you merge it this is resolved, too.

graehl commented:
> > Don't think there's anything I can currently do (please be specific if I'm mistaken, I'm new).
> 
> Rebase YOUR branch to master(then force push to your branch), see 0cc4m's changes, cherry pick 0cc4m(or rebase him changes to your's changes)

Thanks for spelling this out, that was easy - didn't squash so we can keep occam's contrib. separate but it's all rebased and you should see it here.

graehl committed:
SGD op param store weight-decay and not 1-alpha*wd

graehl commented:
> > I'm not aware of anything I can do on my end to get this merged (is someone waiting on me that I'm unaware of?).
> 
> As I said, please use the human-readable parameters, and only the human-readable parameters, as the ones being passed to `ggml_opt_step_sgd`. If you are short on time I can take over and finish up this PR (or someone else can if they want to).

This is fine and done now, but I cannot be confident the vulkan end of things is correct after the change (I just haven't read up on how the vulkan API works, at all).

graehl commented:
> You can change what you want, once things are ready I'll do a proper review of the Vulkan parts and make sure they are okay.

I didn't change anything at all in vulkan - it's all greek to me :) Do take a look. Perhaps the tests weren't really running on vulkan (I had disabled them since I didn't have an impl). The change is that the op params tensor [1] is now sgd.wd instead of 1 - sgd.wd*sgd.alpha. ([0] is just sgd.alpha)

0cc4m commented:
Yeah, no worries. Here's a diff that does that change on the Vulkan shader, and removes two unnecessary preprocessor steps.

```diff
diff --git a/ggml/src/ggml-vulkan/vulkan-shaders/opt_step_sgd.comp b/ggml/src/ggml-vulkan/vulkan-shaders/opt_step_sgd.comp
index 3d5e1d98f..6426dedee 100644
--- a/ggml/src/ggml-vulkan/vulkan-shaders/opt_step_sgd.comp
+++ b/ggml/src/ggml-vulkan/vulkan-shaders/opt_step_sgd.comp
@@ -1,9 +1,6 @@
 #version 450
 
 #include "generic_head.comp"
-#include "types.comp"
-
-#extension GL_EXT_control_flow_attributes : enable
 
 layout(local_size_x = 512, local_size_y = 1, local_size_z = 1) in;
 
@@ -19,7 +16,7 @@ void main() {
     }
 
     const float alpha = data_params[0];
-    const float keep  = data_params[1];
+    const float keep = 1.f - alpha * data_params[1];
 
     data_x[i] = data_x[i] * keep - alpha * data_grad[i];
 }
```

If you apply that the CI should pass again.

JohannesGaessler committed:
minor + cosmetic changes

JohannesGaessler commented:
I made some changes and pushed them to `johannesgaessler/finelayer`. The changes are almost entirely cosmetic, the biggest functional change is that I'm using `ggml_format_name` in `ggml-opt.cpp` instead of manual string manipulation (as I should have done from the start). Training of LLaMA 3.2 1b is currently not working but that is not caused by this PR.

graehl committed:
fix vulkan sgd

graehl commented:
I believe I successfully applied both. If anything else can be done to get this merged let me know.

JohannesGaessler committed:
try CI fix

JohannesGaessler commented:
The build failures are my fault. I don't know why, but for some reason `std::powf` is not available everywhere. I pushed a fix to my repository which just uses `std::pow` again, it's not performance relevant. As I eluded to in an added comment, the code should be refactored to use `logf` and `expf` anyways.

0cc4m reviewed (approved):
The Vulkan part is fine.

JohannesGaessler commented:
Thanks for the work and the persistence, everyone. For bookkeeping I changed the title/commit message to also mention SGD.

WilliamTambellini commented:
Tks @JohannesGaessler
